src/NuGet.Core/NuGet.Build.Tasks/AssignReferencePropertiesTask.cs

```diff
@@ -0,0 +1,127 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Linq;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using NuGet.Commands;
+using NuGet.Frameworks;
+
+namespace NuGet.Build.Tasks
+{
+    public class AssignReferencePropertiesTask : Task
+    {
+        /// <summary>
+        /// The current project's target framework.
+        /// </summary>
+        [Required]
+        public string CurrentProjectTargetFramework { get; set; }
+
+        /// <summary>
+        /// The project references for property lookup.
+        /// </summary>
+        public ITaskItem[] AnnotatedProjectReferences { get; set; }
+
+        /// <summary>
+        /// The project references with assigned properties.
+        /// </summary>
+        [Output]
+        public ITaskItem[] AssignedProjects { get; set; }
+
+        public override bool Execute()
+        {
+            var log = new MSBuildLogger(Log);
+
+            BuildTasksUtility.LogInputParam(log, nameof(CurrentProjectTargetFramework), CurrentProjectTargetFramework);
+
+            if (AnnotatedProjectReferences != null)
+            {
+                log.LogDebug($"(in) AnnotatedProjectReferences '{string.Join(";", AnnotatedProjectReferences.Select(p => p.ItemSpec))}'");
+            }
+            else
+            {
+                return !Log.HasLoggedErrors;
+            }
+
+            var frameworkToMatch = NuGetFramework.Parse(CurrentProjectTargetFramework);
+            if (frameworkToMatch.IsUnsupported)
+            {
+                log.LogError(string.Format(Strings.UnsupportedTargetFramework, CurrentProjectTargetFramework));
+                return false;
+            }
+
+            AssignedProjects = AnnotatedProjectReferences.Select(project => AssignPropertiesForSingleReference(project, frameworkToMatch)).ToArray();
+
+            return !Log.HasLoggedErrors;
+        }
+
+        private ITaskItem AssignPropertiesForSingleReference(ITaskItem project, NuGetFramework currentProjectTargetFramework)
+        {
+            var itemWithProperties = new TaskItem(project);
+
+            var targetFramework = project.GetMetadata("TargetFrameworks");
+
+            if (string.IsNullOrEmpty(targetFramework))
+            {
+                // No target frameworks set, nothing to do.
+                return itemWithProperties;
+            }
+
+            var possibleTargetFrameworks = MSBuildStringUtility.Split(targetFramework);
+            var possibleNuGetFrameworks = possibleTargetFrameworks.Select(ParseFramework).ToList();
+            var nearestNuGetFramework = NuGetFrameworkUtility.GetNearest(possibleTargetFrameworks, currentProjectTargetFramework, NuGetFramework.Parse);
+
+            if (nearestNuGetFramework != null)
+            {
+                if (TryConvertItemMetadataToBool(project, "HasSingleTargetFramework", out var singleTargetFramework) && singleTargetFramework)
```

Comment:
`if (MSBuildStringUtility.IsTrue(project.GetMetadata("HasSingleTargetFramework")))`

I think this could be simplified to the above, it should only need to check if this value is `true` otherwise is false. I might be missing something here though, I don't see where the 2nd bool to see if the metadata was there at all comes into play here.